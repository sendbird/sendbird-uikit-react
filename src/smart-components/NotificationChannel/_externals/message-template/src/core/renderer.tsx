import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
