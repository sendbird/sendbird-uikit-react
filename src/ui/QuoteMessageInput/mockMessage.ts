export const dummyFileMessageImage = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://static.sendbird.com/sample/user_sdk/user_sdk_16.png',
  name: 'Hello i am PDF',
  size: 0,
  type: 'image/pdf',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageVideo = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://sendbird-upload.s3.amazonaws.com/2D7B4CDB-932F-4082-9B09-A1153792DC8D/upload/n/fd4e65eb495f436ab687ba7c5683c922.mp4',
  name: 'https://sendbird-upload.s3.amazonaws.com/2D7B4CDB-932F-4082-9B09-A1153792DC8D/upload/n/fd4e65eb495f436ab687ba7c5683c922.mp4',
  size: 0,
  type: 'video/mp4',
  thumbnails: [{
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 240,
    height: 240
  }, {
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 320,
    height: 320
  }],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageAudio = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  name: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  size: 0,
  type: 'audio/midi',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0,
};

export const dummyFileMessageGif = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://static.sendbird.com/sample/user_sdk/user_sdk_16.png',
  name: 'GIF GIF GIF',
  size: 0,
  type: 'image/gif',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageWithThumbnail = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 751865890,
  messageType: "file",
  channelUrl: "sendbird_group_channel_19238732_8e44958ac5c2fb88203146f42a18a9a4f4472dde",
  data: "",
  customType: "",
  createdAt: 1583104853295,
  updatedAt: 0,
  channelType: "group",
  metaArrays: [],
  reactions: [],
  mentionType: "users",
  mentionedUsers: [],
  url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
  name: "ic-close.svg",
  size: 404,
  type: "image/svg+xml",
  thumbnails: [{
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 240,
    height: 240
  }, {
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 320,
    height: 320
  }],
  sender: {
    nickname: "hoon1000",
    profileUrl: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    userId: "hoon1000",
    connectionStatus: "nonavailable",
    lastSeenAt: 0,
    metaData: {},
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: "1583104837675",
  requireAuth: true,
  requestState: "succeeded",
  requestedMentionUserIds: [],
  errorCode: 0
};

export default { dummyFileMessageAudio, dummyFileMessageVideo, dummyFileMessageImage, dummyFileMessageWithThumbnail }
