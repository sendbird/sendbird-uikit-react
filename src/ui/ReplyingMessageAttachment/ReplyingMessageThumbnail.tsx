import React, { ReactElement } from 'react';
import { FileMessage } from 'sendbird';

import Icon, { IconTypes, IconColors } from '../Icon';
import ImageRenderer from '../ImageRenderer';
import { isAudioMessage, isFileMessage, isThumbnailMessage } from '../../utils';

interface Props {
  message: FileMessage;
}

const componentClassname = 'sendbird-replying_message_attachment__avatar';

export default function ReplyingMessageThumbnail({ message }: Props): ReactElement {
  if (!isFileMessage(message)) {
    return null;
  }

  const thumbnailUrl: string = message.thumbnails && message.thumbnails.length > 0 && message.thumbnails[0].url;
  if (isThumbnailMessage(message) && thumbnailUrl) {
    return (
      <ImageRenderer
        className={componentClassname}
        url={thumbnailUrl}
        alt={message.type}
        width="44px"
        height="44px"
      />
    );
  } else if (isAudioMessage(message)) {
    return (
      <div className={componentClassname}>
        <Icon
          type={IconTypes.FILE_AUDIO}
          fillColor={IconColors.ON_BACKGROUND_2}
          width="24px"
          height="24px"
        />
      </div>
    );
  } else {
    return (
      <div className={componentClassname}>
        <Icon
          type={IconTypes.FILE_DOCUMENT}
          fillColor={IconColors.ON_BACKGROUND_2}
          width="24px"
          height="24px"
        />
      </div>
    );
  }
}
