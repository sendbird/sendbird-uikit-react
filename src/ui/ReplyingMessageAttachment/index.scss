@import '../../styles/variables';

.sendbird-replying_message_attachment {
  position: relative;
  padding: 16px 80px 16px 40px;
  width: 100%;
  height: 76px;
  display: flex;
  flex-direction: row;
  box-sizing: border-box;

  .sendbird-replying_message_attachment__avatar {
    width: 44px;
    min-width: 44px;
    height: 44px;
    min-height: 44px;
    position: relative;
    margin-right: 8px;
    border-radius: 8px;
    display: flex;
    justify-content: center;
    align-items: center;
    @include themed() {
      background-color: t(bg-1);
    }
  }

  .sendbird-replying_message_attachment__body {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    overflow: hidden;

    .sendbird-replying_message_attachment__body__sender-name {
      width: 100%;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      box-sizing: border-box;
      margin-bottom: 8px;
    }
    .sendbird-replying_message_attachment__body__message-content {
      width: 100%;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      box-sizing: border-box;
    }
  }

  .sendbird-replying_message_attachment__close-button {
    position: absolute;
    top: 26px;
    right: 40px;
    &:hover {
      cursor: pointer;
    }
  }
}
