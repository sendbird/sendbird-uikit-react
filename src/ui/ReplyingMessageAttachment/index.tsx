import React, { ReactElement, useContext } from 'react';
import { FileMessage, UserMessage } from 'sendbird';
import {
  CoreMessageType,
  getClassName,
  isFileMessage,
  isGifMessage,
  isImageMessage,
  isThumbnailMessage,
  isUserMessage,
  isVideoMessage,
} from '../../utils';

import Icon, { IconTypes, IconColors } from '../Icon';
import Label, { LabelTypography, LabelColors } from '../Label';
import { LocalizationContext } from '../../lib/LocalizationContext';

import ReplyingMessageThumbnail from './ReplyingMessageThumbnail';
import './index.scss';

interface Props {
  className: string | Array<string>;
  replyingMessage: CoreMessageType;
  onClose?: (message: CoreMessageType) => void;
}

export default function ReplyingMessageAttachment({
  className,
  replyingMessage,
  onClose,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const fileMessage = replyingMessage as FileMessage;
  const sender = (replyingMessage as UserMessage | FileMessage)?.sender;

  return (
    <div className={getClassName(['sendbird-replying_message_attachment', className])}>
      <ReplyingMessageThumbnail message={fileMessage} />
      <div className="sendbird-replying_message_attachment__body">
        <Label
          className="sendbird-replying_message_attachment__body__sender-name"
          type={LabelTypography.CAPTION_1}
          color={LabelColors.ONBACKGROUND_1}
        >
          {`${stringSet.REPLYING_ATTACHMENT__REPLY_TO} ${(sender && sender.nickname) ? sender.nickname : stringSet.NO_NAME}`}
        </Label>
        <Label
          className="sendbird-replying_message_attachment__body__message-content"
          type={LabelTypography.BODY_2}
          color={LabelColors.ONBACKGROUND_3}
        >
          {isImageMessage(fileMessage) && !isGifMessage(fileMessage) && stringSet.REPLYING_ATTACHMENT__FILE_TYPE__IMAGE}
          {isVideoMessage(fileMessage) && stringSet.REPLYING_ATTACHMENT__FILE_TYPE__VIDEO}
          {isGifMessage(fileMessage) && stringSet.REPLYING_ATTACHMENT__FILE_TYPE__GIF}
          {isUserMessage(replyingMessage as UserMessage) && (replyingMessage as UserMessage).message}
          {(isFileMessage(fileMessage) && !isThumbnailMessage(fileMessage)) && fileMessage.name}
        </Label>
      </div>
      <Icon
        className="sendbird-replying_message_attachment__close-button"
        type={IconTypes.CLOSE}
        fillColor={IconColors.ON_BACKGROUND_2}
        width="24px"
        height="24px"
        onClick={() => onClose(replyingMessage)}
      />
    </div>
  );
}
