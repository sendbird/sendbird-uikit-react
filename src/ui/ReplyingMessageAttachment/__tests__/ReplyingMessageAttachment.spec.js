import React from 'react';
import renderer from 'react-test-renderer';

import ReplyingMessageAttachment from "../index";
import { dummyFileMessageAudio } from '../mockMessage.ts';

describe('ReplyingMessageAttachment', () => {
  it('should do a snapshot test of the ReplyingMessageAttachment DOM', function() {
    const text = "example-text";
    const component = renderer.create(
      <ReplyingMessageAttachment replyingMessage={dummyFileMessageAudio} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
