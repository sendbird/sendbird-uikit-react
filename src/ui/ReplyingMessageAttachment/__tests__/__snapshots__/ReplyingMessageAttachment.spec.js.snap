// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReplyingMessageAttachment should do a snapshot test of the ReplyingMessageAttachment DOM 1`] = `
<div
  className="sendbird-replying_message_attachment "
>
  <div
    className="sendbird-replying_message_attachment__avatar"
  >
    <div
      className=" sendbird-icon sendbird-icon-color--on-background-2"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "24px",
          "minHeight": "24px",
          "minWidth": "24px",
          "width": "24px",
        }
      }
      tabIndex="0"
    >
      <test-file-stub />
    </div>
  </div>
  <div
    className="sendbird-replying_message_attachment__body"
  >
    <span
      className="sendbird-replying_message_attachment__body__sender-name sendbird-label sendbird-label--caption-1 sendbird-label--color-onbackground-1"
    >
      Reply to ME
    </span>
    <span
      className="sendbird-replying_message_attachment__body__message-content sendbird-label sendbird-label--body-2 sendbird-label--color-onbackground-3"
    >
      72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3
    </span>
  </div>
  <div
    className="sendbird-replying_message_attachment__close-button sendbird-icon sendbird-icon-color--on-background-2"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    style={
      Object {
        "height": "24px",
        "minHeight": "24px",
        "minWidth": "24px",
        "width": "24px",
      }
    }
    tabIndex="0"
  >
    <test-file-stub />
  </div>
</div>
`;
